File: backend/app/schemas/news.py
from datetime import datetime
from typing import List, Optional, Dict, Any, Union
from pydantic import BaseModel, Field, ConfigDict, field_validator

class NewsItemBase(BaseModel):
    title: str
    url: str
    source: str
    category: str
    summary: Optional[str] = ""
    author: Optional[str] = None
    tags: List[str] = Field(default_factory=list)

    # LLM Synthesis & Curation
    tldr_bullets: List[str] = Field(default_factory=list)
    why_it_matters: Optional[str] = ""
    tech_stack: List[str] = Field(default_factory=list)
    quality_score: float = 6.0
    is_high_signal: bool = False

    raw_score: int = 0
    comments_count: int = 0
    hotness_score: float = 0.0
    published_at: datetime

class NewsItemResponse(NewsItemBase):
    id: int
    created_at: datetime
    dedup_hash: str
    model_config = ConfigDict(from_attributes=True)

class NewsListResponse(BaseModel):
    items: List[NewsItemResponse]
    total: int
    page: int
    size: int
    total_pages: int

class AskQuestionRequest(BaseModel):
    question: str

class AskQuestionResponse(BaseModel):
    news_id: int
    question: str
    answer: str

class BriefingCategoryItem(BaseModel):
    id: Optional[int] = None
    title: str
    url: str
    source: Optional[str] = ""
    why_it_matters: Optional[str] = ""
    summary: Optional[str] = ""
    tags: Optional[List[str]] = []
    category: Optional[str] = ""
    category_name: Optional[str] = ""
    hotness_score: Optional[float] = 0.0

    @field_validator("tags", mode="before")
    @classmethod
    def parse_tags(cls, v):
        if isinstance(v, str):
            return [t.strip() for t in v.split(",") if t.strip()]
        if isinstance(v, list):
            return v
        return []

class BriefingCategorySection(BaseModel):
    id: str
    name: str
    icon: str
    summary: str
    items: List[BriefingCategoryItem] = []

class DailyBriefingResponse(BaseModel):
    title: Optional[str] = "오늘의 AI 브리핑"
    date: str
    headline: str
    overview: Optional[str] = ""
    categories: Optional[List[BriefingCategorySection]] = []
    action_items: Optional[List[str]] = []
    markdown_report: str
    featured_items_count: int
    generated_at: str

class NotionExportRequest(BaseModel):
    api_key: Optional[str] = None
    page_id: Optional[str] = None

class NotionExportResponse(BaseModel):
    success: bool
    message: str
    url: Optional[str] = None

class EmailSendRequest(BaseModel):
    to_email: Optional[str] = None
    subject: Optional[str] = None

class EmailSendResponse(BaseModel):
    success: bool
    message: str

class TestNotionRequest(BaseModel):
    api_key: Optional[str] = None
    page_id: Optional[str] = None

class TestEmailRequest(BaseModel):
    smtp_host: Optional[str] = None
    smtp_port: Optional[int] = 587
    smtp_user: Optional[str] = None
    smtp_password: Optional[str] = None
    use_tls: Optional[bool] = True
    to_email: Optional[str] = None

class BriefingSettingsUpdate(BaseModel):
    notion_api_key: Optional[str] = None
    notion_page_id: Optional[str] = None
    notion_auto_export: Optional[bool] = None
    smtp_host: Optional[str] = None
    smtp_port: Optional[int] = None
    smtp_user: Optional[str] = None
    smtp_password: Optional[str] = None
    smtp_from: Optional[str] = None
    smtp_to: Optional[str] = None
    smtp_use_tls: Optional[bool] = None
    email_auto_send: Optional[bool] = None
    slack_webhook_url: Optional[str] = None
    discord_webhook_url: Optional[str] = None

class ScheduleStatusResponse(BaseModel):
    current_time: str
    schedule_hours: List[int]
    timezone: str
    next_run: str
    minutes_until_next_run: float
    is_running: bool
    last_crawl_time: Optional[str] = None
    last_crawl_status: Optional[str] = "idle"
    recent_logs: List[Dict[str, Any]] = Field(default_factory=list)

class StatsResponse(BaseModel):
    total_news: int
    category_counts: Dict[str, int]
    source_counts: Dict[str, int]
    last_updated: Optional[str] = None

class ManualCollectResponse(BaseModel):
    message: str
    total_collected: int
    total_saved: int
    source_results: Dict[str, Any]

File: frontend/src/types/news.ts
export type CategoryType = 'all' | 'harness' | 'mcp_plugins_skills' | 'agent_tech' | 'ai_news';

export interface NewsItem {
  id: number;
  title: string;
  url: string;
  source: string;
  category: string;
  summary: string;
  author: string | null;
  tags: string[];

  // LLM Synthesis & Curation
  tldr_bullets: string[];
  why_it_matters: string;
  tech_stack: string[];
  quality_score: number;
  is_high_signal: boolean;

  raw_score: number;
  comments_count: number;
  hotness_score: number;
  published_at: string;
  created_at: string;
  dedup_hash: string;
}

export interface NewsListResponse {
  items: NewsItem[];
  total: number;
  page: number;
  size: number;
  total_pages: number;
}

export interface BriefingCategoryItem {
  id?: number;
  title: string;
  url: string;
  source?: string;
  why_it_matters?: string;
  summary?: string;
  tags?: string[];
  category?: string;
  category_name?: string;
  hotness_score?: number;
}

export interface BriefingCategorySection {
  id: string;
  name: string;
  icon: string;
  summary: string;
  items: BriefingCategoryItem[];
}

export interface DailyBriefing {
  title?: string;
  date: string;
  headline: string;
  overview?: string;
  categories?: BriefingCategorySection[];
  action_items?: string[];
  markdown_report: string;
  featured_items_count: number;
  generated_at: string;
}

export interface NotionExportResponse {
  success: boolean;
  message: string;
  url?: string;
}

export interface EmailSendResponse {
  success: boolean;
  message: string;
}

export interface BriefingSettings {
  title: string;
  auto_dispatch_hour: number;
  notion: {
    enabled: boolean;
    has_api_key: boolean;
    masked_api_key: string;
    page_id: string;
    auto_export: boolean;
  };
  email: {
    enabled: boolean;
    smtp_host: string;
    smtp_port: number;
    smtp_user: string;
    has_password: boolean;
    smtp_from: string;
    smtp_to: string;
    use_tls: boolean;
    auto_send: boolean;
  };
  webhooks: {
    slack_webhook_url: string;
    discord_webhook_url: string;
    auto_dispatch_daily_briefing: boolean;
    auto_alert_high_signal: boolean;
  };
}

export interface AskQuestionResponse {
  news_id: number;
  question: string;
  answer: string;
}

export interface CrawlLog {
  id: number;
  source: string;
  items_crawled: number;
  items_saved: number;
  status: string;
  started_at: string | null;
  completed_at: string | null;
}

export interface ScheduleStatus {
  current_time: string;
  schedule_hours: number[];
  timezone: string;
  next_run: string;
  minutes_until_next_run: number;
  is_running: boolean;
  last_crawl_time: string | null;
  last_crawl_status: string;
  recent_logs: CrawlLog[];
}

export interface StatsResponse {
  total_news: number;
  category_counts: Record<string, number>;
  source_counts: Record<string, number>;
  last_updated: string | null;
}
